proposals/0088-enable-core-bpf-programs.md

```diff
@@ -0,0 +1,143 @@
+devnet, then mainnet-beta. This is extremely valuable for ensuring a change has
+been safely integrated and will not have negative effects on mainnet-beta.
+
+This same explicit process can be used to migrate existing native programs to
+their new Core BPF versions.
+
+## Alternatives Considered
+
+An alternative approach to managing upgrades to essential core programs could be
+to make these core BPF programs _upgradeable_ and establish a multi-signature
+upgrade authority. This authority could be comprised of key-holders from each
+validator client: Solana Labs, Jito, Jump, and possibly Syndica (Sig).
+
+However, this approach adds additional layers of complexity around upgrading
+these programs behind a feature-gate.
+
+## New Terminology
+
+- `Core BPF Program`: A **non-upgradeable** BPF program relied on by any part of
+  the Solana stack, including (but not limited to) consensus, transaction
+  processing, voting, staking, and account creation.
+- `Target Program`: The core program intending to be changed by some
+  feature activation.
+- `Source Program`: The modified version of a core program intending to be moved
+  in place of the existing core program.
+
+## Detailed Design
+
+Core BPF programs shall be non-upgradeable BPF programs deployed with
+`BPFLoader2111111111111111111111111111111111`.
+
+Core BPF programs must only be modified through feature gates.
+
+**Upgrading a core BPF program** shall consist of deploying the modified program
+source code to a new arbitrary address as a **non-upgradeable** BPF program and
+creating a feature gate to, using the runtime, move the modified program in
+place of the existing program. 
+
+Note, since this deployed program will be deployed using the non-upgradeable BPF
+loader (`BPFLoader2111111111111111111111111111111111`), it will consist of only
+a program account, with no additional program _data_ account.
+
+This process must adhere to the following steps:
+
+1. Verifiably build the modified program.
+2. Generate a new keypair for the **source** program.
+3. Deploy the program to the **source** address.
+4. Generate a new keypair for the **feature gate**.
+5. Create a new feature gate for replacing the **target** program with the
+   **source** (modified) program.
+6. Follow the existing process for activating features.
+
+An additional optional field could be added to feature gate issues for the
+**target program** being upgraded.
+
+**Migrating a native program** to a core BPF program shall consist of following
+the exact steps outlined above with the modified BPF version of the program.
+This will effectively replace the program at the target address with its core
+BPF implementation.
+
+Some additional checks should be run when doing a migration instead of an
+upgrade, such as:
+
+- Validating the native program's owner is
+  `NativeLoader1111111111111111111111111111111`
+- Ensuring the core BPF program has the correct state
```

Comment:
Sorry, what does this mean?

Reply:
I was originally moving toward some validation/checks that had to do with native program instruction/state, but I don't think it's going to work or worth adding.

This was a straggler from those ideas.